satellite/repair/checker/observer.go

```diff
@@ -0,0 +1,393 @@
+	observer.counter2.Inc(repPartial.monStats.remoteSegmentsOverThreshold[1]) //mon:locked
+	observer.counter3.Inc(repPartial.monStats.remoteSegmentsOverThreshold[2]) //mon:locked
+	observer.counter4.Inc(repPartial.monStats.remoteSegmentsOverThreshold[3]) //mon:locked
+	observer.counter5.Inc(repPartial.monStats.remoteSegmentsOverThreshold[4]) //mon:locked
+
+	return nil
+}
+
+// Finish is called after all segments are processed by all observers.
+func (observer *RangedLoopObserver) Finish(ctx context.Context) error {
+	// remove all segments which were not seen as unhealthy by this checker iteration
+	healthyDeleted, err := observer.repairQueue.Clean(ctx, observer.startTime)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	mon.IntVal("remote_files_checked").Observe(observer.TotalStats.objectsChecked)                               //mon:locked
+	mon.IntVal("remote_segments_checked").Observe(observer.TotalStats.remoteSegmentsChecked)                     //mon:locked
+	mon.IntVal("remote_segments_failed_to_check").Observe(observer.TotalStats.remoteSegmentsFailedToCheck)       //mon:locked
+	mon.IntVal("remote_segments_needing_repair").Observe(observer.TotalStats.remoteSegmentsNeedingRepair)        //mon:locked
+	mon.IntVal("new_remote_segments_needing_repair").Observe(observer.TotalStats.newRemoteSegmentsNeedingRepair) //mon:locked
+	mon.IntVal("remote_segments_lost").Observe(observer.TotalStats.remoteSegmentsLost)                           //mon:locked
+	mon.IntVal("remote_files_lost").Observe(int64(len(observer.TotalStats.objectsLost)))                         //mon:locked
+	mon.IntVal("healthy_segments_removed_from_queue").Observe(healthyDeleted)                                    //mon:locked
+	allUnhealthy := observer.TotalStats.remoteSegmentsNeedingRepair + observer.TotalStats.remoteSegmentsFailedToCheck
+	allChecked := observer.TotalStats.remoteSegmentsChecked
+	allHealthy := allChecked - allUnhealthy
+	mon.FloatVal("remote_segments_healthy_percentage").Observe(100 * float64(allHealthy) / float64(allChecked)) //mon:locked
+
+	return nil
+}
+
+// repairPartial implements the ranged loop Partial interface.
+//
+// architecture: Observer
+type repairPartial struct {
+	repairQueue      *queue.InsertBuffer
+	nodestate        *ReliabilityCache
+	statsCollector   *statsCollector
+	monStats         aggregateStats // TODO(cam): once we verify statsCollector reports data correctly, remove this
+	repairOverrides  RepairOverridesMap
+	nodeFailureRate  float64
+	getNodesEstimate func(ctx context.Context) (int, error)
+	log              *zap.Logger
+	lastStreamID     uuid.UUID
+}
+
+// newRangedLoopCheckerPartial creates new checker partial instance.
+func newRangedLoopCheckerPartial(observer *RangedLoopObserver) rangedloop.Partial {
+	// we can only share thread-safe objects.
+	return &repairPartial{
+		repairQueue:      observer.createInsertBuffer(),
+		nodestate:        observer.nodestate,
+		statsCollector:   newStatsCollector(),
```

Comment:
I don't think we can use the new stats collector with each partial. Each time this is called, it adds the stats to the monitoring chain. This will continously grow the monitoring chain with stale unused stats collectors. This is not a problem with the node tally checker  because it is only initialized once.

Further, I think calling collectAggregates once per partial is going to give confusing metrics that don't represent the whole segment run.